src/main/java/com/direwolf20/buildinggadgets/client/gui/materiallist/MaterialListGUI.java

```diff
@@ -1,217 +1,268 @@
+import net.minecraft.client.gui.screen.Screen;
+import net.minecraft.client.gui.widget.Widget;
+import net.minecraft.client.gui.widget.button.Button;
+import net.minecraft.item.ItemStack;
+import net.minecraft.util.ResourceLocation;
+import net.minecraft.util.text.TranslationTextComponent;
+import net.minecraftforge.common.util.LazyOptional;
+import net.minecraftforge.fml.client.config.GuiUtils;
+
+import java.awt.*;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class MaterialListGUI extends Screen {
+
+    public static int getXForAlignedRight(int right, int width) {
+        return right - width;
+    }
+
+    public static int getXForAlignedCenter(int left, int right, int width) {
+        return left + (right - left) / 2 - width / 2;
+    }
+
+    public static int getYForAlignedCenter(int top, int bottom, int height) {
+        return top + (bottom - top) / 2 - height / 2;
+    }
+
+    public static int getYForAlignedBottom(int bottom, int height) {
+        return bottom - height;
+    }
+
+    public static void renderTextVerticalCenter(String text, int leftX, int top, int bottom, int color) {
+        FontRenderer fontRenderer = Minecraft.getInstance().fontRenderer;
+        int y = getYForAlignedCenter(top, bottom, fontRenderer.FONT_HEIGHT);
+        GlStateManager.enableTexture();
+        fontRenderer.drawString(text, leftX, y, color);
+    }
+
+    public static void renderTextHorizontalRight(String text, int right, int y, int color) {
+        FontRenderer fontRenderer = Minecraft.getInstance().fontRenderer;
+        int x = getXForAlignedRight(right, fontRenderer.getStringWidth(text));
+        GlStateManager.enableTexture();
+        fontRenderer.drawString(text, x, y, color);
+    }
+
+    public static boolean isPointInBox(double x, double y, int bx, int by, int width, int height) {
+        return x >= bx &&
+                y >= by &&
+                x < bx + width &&
+                y < by + height;
+    }
+
+    public static final int BUTTON_HEIGHT = 20;
+    public static final int BUTTONS_PADDING = 4;
+
+    public static final ResourceLocation BACKGROUND_TEXTURE = new ResourceLocation(Reference.MODID, "textures/gui/material_list.png");
+    public static final int BACKGROUND_WIDTH = 256;
+    public static final int BACKGROUND_HEIGHT = 200;
+    public static final int BORDER_SIZE = 4;
+
+    public static final int WINDOW_WIDTH = BACKGROUND_WIDTH - BORDER_SIZE * 2;
+    public static final int WINDOW_HEIGHT = BACKGROUND_HEIGHT - BORDER_SIZE * 2;
+
+     // Item name (localized)
+     // Item count
+    public static final String PATTERN_SIMPLE = "%s: %d";
+     // Item name (localized)
+     // Item count
+     // Item registry name
+     // Formatted stack count, e.g. 5x64+2
+    public static final String PATTERN_DETAILED = "%s: %d (%s, %s)";
+
+    private int backgroundX;
+    private int backgroundY;
+    private ItemStack item;
+    private ITemplate cachedTemplate;
```

Comment:
You should not cache the Template - when you first query it, it may not be present on the client yet, as it might need to be fetched from the Server first.

Do you need a Handle System (for example via a Supplier) so that you can cache that?

Reply:
Removed cache, because after update I realize it will only be queried once.